faster line iterator

pascalkuthe opened the pull request:
~~Note: This is a very early draft and still contains known issues. CI failures are known and expected (multiple tests relating to the reverse iterator still fail).
It's just posted here to allow comparison with the approach in #69 .~~

This PR is now mostly ready for review, although I still want to add some more comments and clean up the code a bit. But this PR works now so reviewing makes sense now. I have dropped the unrelated commits and removed debug prints.

This implements a line iterator using a custom chunks iterator that tracks the lowest common parent across line boundries to avoid a binary search in `RopeSlice::new`.

The results are pretty spectacular. On synthetic benchmarks it outperforms the old implementation by a factor between 3x and 9x.
See the run of the builtin benchmarks below (this is the wrong way around, I benchmarked the new implementation first and then the old one):
![image](https://user-images.githubusercontent.com/61850714/200080388-9756a8b4-f8f6-45f7-9972-9b60bd1ed3b7.png)

I have tested the iterator in helix with helix-editor/helix#4457, reloading a 1GB file improves the **total reload time** (excluding IO) from 14 seconds to 7 second.

cessen commented:
This is definitely the approach I wanted to take with this.  Indeed, it's subtle to get right, which is why I put it off for so long, ha ha.  Thanks for taking this on!  I'll take a look at the actual code when my brain is functioning properly again.

cessen commented:
Ah... maybe I'm forgetting how github works, but it looks like the commits from your other PRs are rolled in with this one?  Could you keep your PRs on separate branches so it's easier to review?  (And also easier to accept some PRs without having to accept others too early.)

pascalkuthe commented:
Hey @cessen, thanks for all the feedback. Please take the time to recover, I hope you get well soon! Regarding the extra commits you are totally right about that. This PR is not ready for review yet and not at all in a state I eould usually post it. It just wanted to post mybstate for #69. here are still known

archseer commented:
Should we also include https://github.com/cessen/ropey/pull/69/commits/159459652c2e33c2423e7f6da39d248beec13c85?

pascalkuthe commented:
> Should we also include [1594596](https://github.com/cessen/ropey/commit/159459652c2e33c2423e7f6da39d248beec13c85)?

I would hope that this optimizes to the same code but I will check the assembly later to make sure. If the compiler can indeed not optimize this then including it might add a small performance win

pascalkuthe commented:
BTW the current implementation does not yet use SIMD for finding the next newline within a byte because the functions available in byte_idx do not do what I need. I already have a SIMD version of `find_next_line_break` (same as line_to_byte_idx(text, 1) but returns None if no line break is found to disambiguate a newline char at the very end from no newline char) implemented and the reverse iteration function should be decently easy as well. From testing I did locally this has the potential to improve performance even further as most time is actually spent on finding the newlines within chunks and the SIMD versionof those functions is at lest 2x faster (for lines linger then 16 bytes)

Should I prepare a PR to str_indices for that and block this PR on that or should that be a follow-up PR?

archseer commented:
Maybe open a draft PR targetting this PR. Github will automatically switch it to target `master` if this one lands first.

cessen commented on src/str_utils.rs:
```diff
@@ -59,6 +59,25 @@ pub(crate) fn utf16_code_unit_to_char_idx(text: &str, utf16_idx: usize) -> usize
     str_indices::chars::from_byte_idx(text, str_indices::utf16::to_byte_idx(text, utf16_idx))
 }
 
+/// Returns the byte position just after the fist line break
+/// in `text`, or `None` if there is no line break.
+/// This function currrently does not utilise SIMD
+pub fn next_line_break(text: &str) -> Option<usize> {
```
You should be able to implement this in terms of `lines::to_byte_idx()`, I think.  Basically just `to_byte_idx(text, 1)` with an additional check if you get a one-past-the-end result.

Your implementation here is also incorrect with respect to feature flags: the recognized line endings are configurable via feature flag, whereas your implementation here is fixed to always recognize all line endings regardless of feature flags.  So implementing in terms of `lines::to_byte_idx()` would fix that as well (the correct `lines` module is conditionally imported in this file already).

pascalkuthe replied:
The difference is a lot more subtle, I used to actually do exactly that but it caused subtly broken results, because you can not differentiate between a line terminator at the end of the line and no line terminator. For example `to_byte_idx("fo\n", 1)` and  `to_byte_idx("foo", 1)` both output  `3`. Correctly handling this would require checking for a line ending again which is really cumbersome and also slow (remeber that this occurs at the end of every chunk not just EOF). Furthermore the case of only looking for the first newline terminator can be easier optimized because it requires fewer checks so its even faster for small lines (which are fairly common). I will try to put together a draft PR to `str_indicies` tomorrow I sadly had a bit much on my plate today. I would also like to SIMD accelerate the reverse case which is not possible to implement with any function is `str_indicies` at all (and was already custom implemented for the old lines iterator). It won't be a big PR at all (like 30 lines for the forward case) because the core SIMD function for counting lines can be reused.

cessen replied:
> because you can not differentiate between a line terminator at the end of the line and no line terminator.

That's what I meant by "with an additional check if you get a one-past-the-end result".  Basically, what I'm thinking is this (untested):

```rust
pub fn next_line_break(text: &str) -> Option<usize> {
    let byte_idx = lines::to_byte_idx(text, 1);
    if byte_idx == text.len() && !ends_with_line_break(text) {
        None
    } else {
        Some(byte_idx)
    }
}
```

(Note that `ends_with_line_break()` is already implemented, so you don't need to re-implement.)

> Correctly handling this would require checking for a line ending again which is really cumbersome and also slow

The check only happens when you reach the end of the text, so I suspect the amortized cost will be small.

> Furthermore the case of only looking for the first newline terminator can be easier optimized because it requires fewer checks

While that's true, I also think there's a balance to be struck between squeezing for performance and having a smaller code base with a smaller API surface area.  If it turns out it's a large performance boost, then it may well be worth it, and I'll certainly consider merging the PR.  But be prepared for a possible rejection as well.

> I would also like to SIMD accelerate the reverse case which is not possible to implement with any function is str_indicies at all

This, on the other hand, is something I would love to have in `str_indices`.  For the `chars` and `utf16` modules as well (though you don't have to tackle those too, just thinking out loud).  But I'd prefer that it be the generalized version, as in `fn to_byte_idx_rev(text: &str, line_idx_rev: usize)` rather than a narrow "prev" method, as per my comments above about the forward case.

pascalkuthe replied:
I have switched back to the `line_to_byte_idx(_, 1)`

cessen commented on tests/proptest_tests.rs:
```diff
@@ -717,6 +717,10 @@ proptest! {
     fn pt_lines_iter_exact_01(idx in 1024usize..(CHAR_LEN - 1024)) {
         let r = Rope::from_str(TEXT);
         let s = r.slice(idx..(idx + 373));
+        // newline terminator must never be split
+        if s.char(372) == '\r'{
+            return Ok(())
+        }
```
This code addition looks incorrect to me.  There's nothing wrong with slicing a rope in the middle of a CRLF pair.  Moreover, stand-alone CR's are perfectly okay and can occur.  So this is a case that actually should be tested, not skipped.

pascalkuthe replied:
Sorry I meant to add a comment and forgot. I am not sure if I ran into a bug in ropey here or how to handle this.
Without this line, this tests failed because the `size_hint` of the lines iterator under reported the line by one (so the iterator did one extra iteration and caused an iteger underflow).
However the size_hint simply reports the number of lines returned by `slice.len_lines()` (and that is check with an assert in this test). So the problem is that slicing in the middle of a CRLF pair causes `RopeSlice::len_lines()` to return the wrong value.
I assumed that you were not supposed to slice in the middle of a CRLF pair (since chunks are explicitly not split in the middle of CRLF pairs). If that is not a case I guess its an (unrelated) bug

cessen replied:
> So the problem is that slicing in the middle of a CRLF pair causes RopeSlice::len_lines() to return the wrong value.

Ah!  Thanks for the bug catch.  Indeed, slicing fails to account for splitting a CRLF pair at its end.  Working on a fix now.

cessen replied:
Fixed in 8699de0908b3853431e2dcac6eb70faad7f326d0

cessen replied:
Also, the additional unit tests I added in the fix revealed that the current lines iterator doesn't handle this case properly.  So your PR here will not only make `Lines` way faster, but also more correct! :-)

pascalkuthe replied:
your tests exposed a couple extra edgecases so its really great we cought this :) I have pushed fixes (see comment below)

cessen commented:
Those results look great!  Awesome work.

I still don't have the energy for a full code review yet, but I left a couple of notes that I think can be addressed already.

cessen commented:
> Should I prepare a PR to str_indices for that and block this PR on that or should that be a follow-up PR?

I don't think any changes to str_indices are necessary.  I made an in-line review comment about how to implement your function in terms of `lines::to_byte_idx()`.

pascalkuthe commented:
> Those results look great! Awesome work.
> 
> I still don't have the energy for a full code review yet, but I left a couple of notes that I think can be addressed already.

Thanks for your comments! I replied to these, both of those are on me.
I anticipated these exact review comments (were my first thoughts aswell) and wanted to add review comments myself to preempt them but I ended up forgetting about it. Sorry about that!
Please take your time to recover, I hope you feel better soon! Health always comes first.

pascalkuthe commented:
@cessen I am not sure why the three tests are failing in the miri CI. They work fine in the normal CI/locally? Maybe some SIMD free fallback function used by MIRI behaves differently then the normal function? It doesn't seem related to this PR

cessen commented on src/iter.rs:
```diff
@@ -1125,43 +767,468 @@ impl<'a> Chunks<'a> {
+
+//==========================================================
+
+/// An iterator over a `Rope`'s contiguous `str` chunks.
+///
+/// Internally, each `Rope` stores text as a segemented collection of utf8
+/// strings. This iterator iterates over those segments, returning a
+/// `&str` slice for each one.  It is useful for situations such as:
+///
+/// - Writing a rope's utf8 text data to disk (but see
+///   [`write_to()`](crate::rope::Rope::write_to) for a convenience function that does this
+///   for casual use-cases).
+/// - Streaming a rope's text data somewhere.
+/// - Saving a rope to a non-utf8 encoding, doing the encoding conversion
+///   incrementally as you go.
+/// - Writing custom iterators over a rope's text data.
+///
+/// There are precisely two guarantees about the yielded chunks:
+///
+/// - All non-empty chunks are yielded, and they are yielded in order.
+/// - CRLF pairs are never split across chunks.
+///
+/// There are no guarantees about the size of yielded chunks, and except for
+/// CRLF pairs and being valid `str` slices there are no guarantees about
+/// where the chunks are split.  For example, they may be zero-sized, they
+/// don't necessarily align with line breaks, etc.
+#[derive(Debug, Clone)]
+pub struct Lines<'a> {
```
It looks like you accidentally copy-pasted the `Chunks` iterator documentation, overwritten the correct `Lines` iterator documentation.

pascalkuthe replied:
Yes you are right, thats the case for all functions. I started off with the chunks iterator as a blueprint andd totally forgot about fixing that! I am not home right now but I will fix this first thing tomorrow morning

cessen replied:
It also looks like you moved the `Lines` iterator to be below the `Chunks` iterator, which seems to be adding to the noisiness of the diff on Github.  If you could move it back (the order should be: `Bytes`, `Chars`, `Lines`, `Chunks`) I'd really appreciate it.  I don't really care about the order per se, but it would make this already-challenging PR a little easier to review.

(And just in general, trying to minimize irrelevant changes in PRs is usually appreciated by reviewers, regardless of project.)

cessen replied:
> I am not home right now but I will fix this first thing tomorrow morning

No rush!  This PR is going to take a while for me to fully review, because of how involved it is.

pascalkuthe replied:
I have fixed the documentation (and also updated it to reflect the now improved time complexity) and moved the `Lines` iterator back to the previous position. A git diff will not be all that useful tough for this PR because ther is basically no carry-over from the old implementation (the `Lines` iterator is closer now to the `Chunks` iterator then the old `Lines` iterator) so I woud suggest just looking at the implementation instead of the diff while reviewing

cessen replied:
Yeah, I don't expect the diff to be *especially* useful.  But when scrolling through to make comments, I was having to scroll all the way past the "added" Chunks iterator (which was actually just moved).  I also don't expect the "light" variant of the code to really be meaningfully different, and this makes that easier to look at as well.

cessen commented:
> Because this PR is already quite big (and a large performance win in every situation) I would like to outscope the SIMD implementation for the reverse iteration for now.

Yeah, I think that's a good call.  Honestly, it's not totally clear to me that it's worth bothering with at all.  Because although reverse `Char` and `Byte` iterators have some common performance-critical use-cases, I struggle to think of similar cases for reverse `Lines` iterators.  So it might be one of those "wait until someone has a use case" things.

> It seems the same MIRI tests fail on master too so its probably not related to this PR.

Yeah, I doubt it's related to this PR.  Feel free to ignore them.  I'll look into and fix them before the next release.

cessen commented on src/iter.rs:
```diff
@@ -1125,43 +767,468 @@ impl<'a> Chunks<'a> {
+        )
+    }
+
+    pub(crate) fn new_with_range(
+        node: &Arc<Node>,
+        byte_idx_range: (usize, usize),
+        line_idx_range: (usize, usize),
+    ) -> Lines {
+        Lines::new_with_range_at_impl::<false>(node, 0, byte_idx_range, line_idx_range)
+    }
+
+    pub(crate) fn new_with_range_at(
+        node: &Arc<Node>,
+        at_line: usize,
+        byte_idx_range: (usize, usize),
+        line_idx_range: (usize, usize),
+    ) -> Lines {
+        Lines::new_with_range_at_impl::<true>(node, at_line, byte_idx_range, line_idx_range)
+    }
```
I may be misunderstanding the code here, but it looks to me like `new_with_range()` and `new_with_range_at()` are inconsistent with each other.  The following two calls should be identical:

```rust
Lines::new_with_range(node, byte_range, line_range);
Lines::new_with_range_at(node, 0, byte_range, line_range);
```

But it looks like they're different.  Specifically, the former calls `Lines::new_with_range_at_impl()` with a `false` const generic parameter, and the latter with `true`.

I'm also a little nervous about that const generic parameter existing at all: I would expect it to either always be true or always be false.  I haven't looked at the code closely enough yet, though, so I could certainly be wrong about that.  But it smells a little to me.

pascalkuthe replied:
Ok this is definitly subtle but its the most elegant way to implement it I could come up with. I will try to explain:

First a small nitpick about your comment so it doesn't get confusing later: 
The two iterators you posted are not always equivalent. They are only equivalent when you are iterating a full `Rope` (or a `RopeSlice` that happens to start within the first line of its top `Node`). However these two iterators are always equal if you use `line_range.0` instead of `0`:

``` rust
Lines::new_with_range(node, byte_range, line_range) 
Lines::new_with_range_at(node, line_range.0, byte_range, line_range)
```

Its trivial to see that my code fulfill that criteria, because if you look into the `new_with_range_at_impl` function it actually special cases `line == line_range.0` to always return `Lines::new_with_range(node, byte_range, line_range)` right at the start of the function.
The reason this special case is required is because the first line might start before the start of the `RopeSlice` (so before `byte_range.0`). That means the line start is potentially in a different chunk then the position you are looking for. To find the correct chunk you need to perform a tree search for the byte position `byte_range.0` instead of the line `line_range.0`.

So really the const generic indicates:
* `false` => start the iterator at the first line
* `true` => start the iterator at the specified line (with a dynamic check if this line is the first line to switch to the other implementation in that case)

This special case might seem odd but I will try to explain my reasoning a bit more:
A similar problem occurs at the end of the `line_range` (so `line >= line_range.1`). I handle this special case as follows:
``` rust
let mut res = Lines::new_with_range_at(node, line_range.1 - 1, byte_range, line_range);
res.next();
res
```
This reuses the logic I already have inside the line iterator to correctly handle the end of the last line and therefore doesn't require yet another specical cased implementation of the `new` function.

You might considered doing the inverse for the start of the line:
``` rust
let mut res = Lines::new_with_range_at(node, line_range.0 + 1, byte_range, line_range);
res.prev();
res
```

However this doesn't work when there is only a single line (because then these two cases just endlessly recuse into each other) so you need to handle one of them as a special case. 
Its admittedly somewhat of arbitrary which one of those two you special case but there are a couple reasons I special cased the first line instead of the last line:
* the implementation is slightly easier/more natural
* the special case is slightly faster (because it does not involve stepping the iterator an additional time) and constructing an iterator at the first line should be way more common (just `lines`) then constructing an iterator at the last line

I think its a bit easier to think about `new_with_range` and `new_with_range_at` as two seperate functions (where `new_with_range_at` will call `new_with_range` for `line == line_range.0`).
These two functions are just very similar so I used the const generic here to share most of the implementation.

cessen replied:
> First a small nitpick about your comment so it doesn't get confusing later:

Ah, right.  Thanks for pointing that out!

> The reason this special case is required is because the first line might start before the start of the RopeSlice (so before byte_range.0). That means the line start is potentially in a different chunk then the position you are looking for. To find the correct chunk you need to perform a tree search for the byte position byte_range.0 instead of the line line_range.0.

I'll keep an eye out for this when I take a closer look at the code.  But at first blush, I think I would have implemented the tree traversal code to account for both the line *and* byte starts, diving into the the furthest-forward of the two as it goes.  Then (unless I'm still missing something, which is possible) there wouldn't need to be any special cases.

> A similar problem occurs at the end of the line_range (so line >= line_range.1). I handle this special case as follows:

I guess part of what's throwing me off here is that it's not clear to me why these cases aren't handled in the traversal/stack-building code itself.  It *feels* to me like there's a simpler implementation struggling to emerge here, but we haven't quite hit it yet.

Partly this is from experiences with my own code in the past: special cases that look a lot like this have very often indicated that my perspective on the problem is a little off.

But to be clear, that doesn't mean the code is incorrect.  Not at all.  Just that a more straightforward and also correct alternative implementation may be possible.

Having said all of that, I don't mean to block this PR on stuff like that.  As long as it's correct and I can follow it, let's get this merged.  Especially considering that there's already a lot of "too complex" code in Ropey already, which is my own doing.  Ha ha. :-)  I can always revisit the code and try to simplify it myself in the future.

pascalkuthe replied:
I thaught about your suggestion regarding doing the tree sesach for lines/bytes all at once but my intuition is that while its theoretically possible it would involve negative indexing (because you always have to keep track of your position relatively to the current node) and a bunch of extra conditions to handle that correctly so the sketch I have in my mind right now is actually more complex then the current implementation.

I can experiment a bit with it (the ideas in my mind are often wrong :D) and if I find s better solution in the future I might do a follow-up PR if you can live with the current implementation for now. This function took a while to to get right (edgecases like a rope slice split inside a CRLF must be considered correctlly) and so overall improving this would take a while

pascalkuthe commented:
There was another problem with CI where `medium.txt` was removed which made the test I added for #67 fail.
I think I might have just forgotten to stage that file  in that PR. I added a commit with that file but feel free to just fix that straight in master. It doesn't warrent another PR I think.

cessen commented on src/iter.rs:
```diff
@@ -555,95 +551,195 @@ impl<'a> ExactSizeIterator for Chars<'a> {}
+                return res;
+            }
+        }
+        // For convenience/readability.
```
I believe in the `Chunks` iterator this comment accompanies the declaration of the `start_byte` and `end_byte` variables.

cessen commented on src/iter.rs:
```diff
@@ -555,95 +551,195 @@ impl<'a> ExactSizeIterator for Chars<'a> {}
+        debug_assert!(node.is_char_boundary(start_byte));
+        debug_assert!(node.is_char_boundary(end_byte));
+
+        // curretly cut off the first line by handeling the same as a full slice
```
Typo, capitalization, and period.

pascalkuthe commented:
Sorry I meant to fix your comments today but totally forgot about it. Its pretty late now here but I will do it tomorrow moring!

pascalkuthe commented:
I fixed the mistakes in the comments and added a ton more. In the process I went trough the entire code again and found two unnecessary branches in the `new_with_range_at` function. These branches had simply no effect anymore (they were only needed in previous versions of the iterator). This made the code simpler.

I think/hope that the added comments make this PR a easier to review for you.
I would recommend starting with the `next_impl` function. The `prev_impl` function is a bit more tricky and I think the context of the constructor isn't really required for reviewing the `next_impl` function.

cessen commented:
Thanks @pascalkuthe!  I'll try to get to this as soon as I can.  A somewhat more urgent task came up in another project, but I think that will probably wrap up within a week.  Let me know if this gets urgent, and I'll bump up the priority.

pascalkuthe commented:
I think @archseer was planning to do a new helix release at the end of this month and wanted to include my two PRs that are blocked on a new ropey version (helix-editor/helix#3890 and helix-editor/helix#4457). It would be awesome if that would workout. I think we could still manage that if we continue next week with the review and nothing mayor turns up but I of-course understand that other things have to take priority @cessen.

cessen commented:
Okay, so things have gotten *even more* hectic.  Long story short: I'm working as the VFX supervisor on a film shoot this week (Mon-Fri).  I thought I was going to have time in the evenings to work on the other task I mentioned, but that mostly doesn't seem to be the case.  So that other task (which also has a deadline of end of the month) is *also* getting postponed, which would push out the full review of this PR even further.

I don't think it makes sense to keep you guys waiting on me, particularly if you want to make a release by the end of the month.  So I think what I might do is this:

1. Do a cursory review, just to make sure nothing obviously wrong jumps out.
2. Merge the PR.
3. Publish a pre-release to crates.io that you guys can use for the upcoming Helix release.
4. When I get the time, review the code properly and just make any changes I'd like to see myself.  And then make a proper release.

Does that sound reasonable?  Basically, I don't want to block you guys, and I definitely want to merge this PR.  But I also don't want to make a "proper" release without thoroughly reviewing the code first.

pascalkuthe commented:
@cessen sorry for the ping. Do you think you could take a look at this soon? The end of November is only a week away so there is not that much time left to land the PRs blocked on this for the next helix release. Quite a few people have been daily driving these changes in helix (by using my PR based on this as their daily driver). In that PR a line iteration is performed on every keypress (and incorrect iterations is imminently apparent). So it's been fuzzed quite a bit.  That together with the very detailed testsuite gives me a lot of confidence in the implementation.

cessen commented:
Hi @pascalkuthe,

Sorry for the silence.  I'll make sure to get to this by at least end of day Sunday (possibly sooner).

cessen commented:
Thanks for your patience!  November ended up being super busy for me, so I wasn't able to get to things as promptly as I prefer.

I've done a cursory review, and although there are a few things I'd like to look into a bit deeper, there's nothing that jumps out to me as incorrect.  Combined with the testing and real-world usage your discussed, I'm more than happy to make an alpha release out of this for Helix to depend on.  And then I can dive deeper a bit more at my leisure over the next month or so to get things ready for a proper release.

Thanks a bunch for your work on this!  It's been a wart in Ropey for a while, so it's awesome to finally get it addressed.
